Doc generation and benchmark modules (#4151)
SpikeInterface/spikeinterface: doc/modules/sortingcomponents.rst

    peak_dtype = [('sample_index', 'int64'), ('channel_index', 'int64'), ('amplitude', 'float64'), ('segment_index', 'int64')]

2 main different methods are available with the :code:`method` argument:

* **'locally_exclusive'** (requires :code:`numba`): peaks on neighboring channels within a certain radius are excluded (not counted multiple times)
* **'matched_filtering'** (requires :code:`numba`): 

Other variant are also implemented (but less tested or not so util):

> * **'by_channel'** (default): peaks are detected separately for each channel, this should be used in high density probe layout.

Author: ```suggestion
* **'by_channel'** : peaks are detected separately for each channel, this should be used in high density probe layout.
```

     * **'z'** is orthogonal to the probe plane

Node pipelines
--------------

Both :py:func:`~spikeinterface.sortingcomponents.peak_detection.detect_peaks()` and
:py:func:`~spikeinterface.sortingcomponents.peak_localization.localize_peaks()` need to walk throughout the entire

> recording traces. So this make reading traces and computing the preprocessing twice : this can be very slow!
> Luckily, there is an internal machinery to avoid the 2 times traces reading : :py:func:`~spikeinterface.core.run_node_pipeline()`

Reviewer: ```suggestion
recording traces, duplicating the reading of the traces from disk and applying the preprocessing.
This can be very slow!

Luckily, there is an internal machinery to avoid the multiple passes over the traces: the :py:func:`~spikeinterface.core.run_node_pipeline()` function.
```

> Hopefully, there is an internal machinery to avoid the 2 times traces reading : :py:func:`~spikeinterface.core.run_node_pipeline()`

Author: of course

The *node pipeline* is an API that runs user-selected *nodes* in parallel on all traces' chunks and performs computations like
**peak detection**, **peak localization**, **svd featuring**, ...

Here is a small example that does peak detection and localization at once.

> In the following, please note that there is a middle node that does not output
> final results : the local waveforms extractor.

Reviewer: ```suggestion
In the following, please note that there is an intermediate node, the `ExtractDenseWaveforms` node, that does not output final results (notice the `return_output=False`), but is needed to extract waveforms for the localization node.
```

Implemented methods are the following:

>   * **uniform'**

Reviewer: ```suggestion
  * **'uniform'**
```

a label for every peak.

Some methods have been implemented with various ideas in mind. We really hope that this list will be extended 
soon by tallented people willing to improve. This is a crucial and not well totally resolved step.

  * **'iterative-hdbscan'** : method used in spkyking-circus2. This performs local hdbscan clusetrings on
     svd waveforms features.
  * **'iterative-isosplit'** :  method used in tridesclous2. This performs local isosplit clusetrings on
     svd waveforms features.
  * **'hdbscan-positions'** : This performs a hdbscan clusetring based on the localizations of the spikes.
    This mimic the herdingspikes approach : make the clustering on spike position only but more flexible
    because more localization methods are availables.
  * **'random-projections'** : attempt to make the feature from waveforms with random projections instead of the
    good-old-scool-pca.
  * **'graph-clustering'** : attempt to resolved the clusetring globally and not locally. This construct a global
    but sparse distance matrix between all spikes. Can be slow. Then it perform 'classical' algos on

>     graph (Louvain, Leiden or even HDBSCAN). Promising method but not as afficient as the 'iterative-isosplit' or

Author: yes I think.

    'iterative-hdbscan'.

.. code-block:: python

  from spikeinterface.sortingcomponents.peak_detection import detect_peaks
  peaks = detect_peaks(recording, ...) # as in above example

  from spikeinterface.sortingcomponents.clustering import find_clusters_from_peaks
  labels, peak_labels = find_clusters_from_peaks(recording=recording, peaks=peaks, method="iterative-isosplit")

* **labels** : contains all possible labels (aka unit_ids)
* **peak_labels** : vector with the same size as peaks containing the label for each peak

Extract SVD from peaks
----------------------

> Importantly many clustering function are using internally the 

Reviewer 2: ```suggestion
Importantly many clustering functions internally use the
```

:py:func:`~spikeinterface.sortingcomponents.clusetring.extract_peaks_svd.extract_peaks_svd()`.

> This run a **node pipeline** that extract on a selected peaks set, etxract waveforms, sparsify then, and compress

Reviewer 2: ```suggestion
This runs a **node pipeline** that extracts on a selected peaks set, extracts waveforms, sparsifies them, and compresses
```

> then on time axis using **svd**.

Reviewer 2: ```suggestion
them on time axis using **svd**.
```

> This runs a **node pipeline** that extracts on a selected peaks set, extracts waveforms, sparsifies them, and compresses
> them on the time axis using **svd**.

Reviewer: ```suggestion
This runs a **node pipeline** on a selected peaks set that extracts waveforms, sparsifies them, and compresses
them on the time axis using **svd**.
```

> them on time axis using **svd**.

Reviewer 2: ```suggestion
them on the time axis using **svd**.
```

At the moment, there are five methods implemented:

>   * **'nearest'**: a simple implemenation which is more or less a np.argmin distance for the spike waveforms against all templates.

Reviewer 2: ```suggestion
  * **'nearest'**: a simple implementation which is more or less a np.argmin distance for the spike waveforms against all templates.
```

>   * **'nearest-svd'**: a smarter implemenation than 'nearest' using svd compression and spatial sparsity.

Reviewer 2: ```suggestion
  * **'nearest-svd'**: a smarter implementation than 'nearest' using svd compression and spatial sparsity.
```

>   * **'tdc-peeler'**: a simple idea similar to 'nearest'. Perform nearest on local detected peaks, fit the amplitude and

Reviewer 2: ```suggestion
  * **'tdc-peeler'**: a simple idea similar to 'nearest'. Perform nearest on locally detected peaks, fit the amplitudes and
```

>     remove it from tarces. Then re-run on residual. A bit naive but this is very fast.

Reviewer 2: ```suggestion
    remove them from the traces. Then re-run on residual. A bit naive but this is very fast.
```

>   * **'circus-omp'**: a more serious implemenation orthogonal template matching. This internally make a convolution

Reviewer 2: ```suggestion
  * **'circus-omp'**: a more serious implementation orthogonal template matching. This internally make a convolution
```

    of traces with all templates with some svd decomposition tricks to be faster. This is quite accurate but
    need lots of memory.

>   * **'wobble'**: this is a re implemenation of the yass template matching codes. Finally, very similar to 'circus-omp'.

Reviewer 2: ```suggestion
  * **'wobble'**: this is a re-implementation of the yass template matching code. Also very similar to 'circus-omp'.
```
